Airflow + SageMaker guide (#225)
astronomer/airflow-guides: guides/airflow-sagemaker.md

This use case assumes that the SageMaker model already exists and is not being managed by Airflow. For example, you might use a SageMaker notebook to do exploratory data analysis and develop and deploy your model, and then use Airflow to execute that model as needed. For this example, we used the [SageMaker notebooks tutorial](https://docs.aws.amazon.com/sagemaker/latest/dg/gs-console.html) which creates an XGBoost model from a census dataset.

For our pipeline, we take data from a local CSV and upload it to S3. Then we make predictions on that data with the SageMaker model by submitting a [SageMaker batch transform job](https://docs.aws.amazon.com/sagemaker/latest/dg/how-it-works-batch.html). Batch transforms are useful when you need to run inferences on large datasets, and don't have a persistent endpoint hosted for your model. They require input data to be provided in S3, and will save output results to an S3 path provided. For the last step in our pipeline, we load the results CSV into Redshift.

To implement this pipeline we have the following steps.:

> - Load the data from local storage (`include/`) into S3 using the `PythonOperator` with `S3Hook`. 

Author: Yep, good call!

    )

    results_to_redshift = S3ToRedshiftOperator(
            task_id='save_results',
            aws_conn_id='aws-sagemaker',
            s3_bucket=s3_bucket,
            s3_key=output_s3_key,
            schema="PUBLIC",
            table="results",
            copy_options=['csv'],
        )

    upload_data >> predict >> results_to_redshift
```

Note that every operator in this DAG utilizes an `aws_conn_id`. For this example, we created an IAM user that has access to all relevant resources, and created the Airflow connection using an access key and token as the Login/Password. If you go this route, the SageMaker operators require a region to be specified as an Extra. The connection should look like this:

![SageMaker Connection](https://assets2.astronomer.io/main/guides/sagemaker/sagemaker_connection.png)

Depending on your Airflow environment, you could also have your Airflow deployment assume an IAM role that has access to those resources.

Now if we run this DAG and go to our SageMaker dashboard, we should see the job show up under Batch transform jobs.

![Transform Job](https://assets2.astronomer.io/main/guides/sagemaker/sagemaker_transform_job.png)

> The output of the transform job will then be uploaded to Redshift, where we can easily access results. This is a simple form of this use case, but you can imagine that the Airflow DAG could easily be extended to acquire data from an external system, clean or pre-process the data if needed, and complete any other tasks necessary to publish the results after saving them to Redshift.

Reviewer: ```suggestion
The output of the transform job will then be uploaded to Redshift, where we can easily access results. This is a simple form of this use case, but you can easily extend this DAG to acquire data from an external system, clean or pre-process the data if needed, and complete any other tasks necessary to publish the results after saving them to Redshift.
```

## Use Case 2: Orchestrate a Full ML Pipeline

Our second use case is using Airflow to orchestrate an end-to-end ML pipeline in SageMaker. We use Airflow to acquire and pre-process the data, train a model, create a model from the training results, and evaluate the model on test data using a batch transform job. This use case is relevant if you want to automate the model training, testing, and deployment parts of your ML pipeline.

For this example, we use the [Iris dataset](https://archive.ics.uci.edu/ml/datasets/iris), and train a built-in SageMaker K-Nearest Neighbors (KNN) model. The general steps in the DAG are:

1. Using a `PythonOperator`, grab the data from the API, complete some pre-processing so the data is compliant with KNN requirements, split into train and test sets, and save them to S3 using the `S3Hook`.

> 2. Train the KNN algorithm on the data using the `SageMakerTrainingOperator`. The configuration for this operator requires information about the algorithm, hyper parameters, the input and output data configuration, resource specifications for the machine running the training job, and the Role ARN for execution. For more information about submitting a training job, check out the API documentation [here](https://docs.aws.amazon.com/sagemaker/latest/APIReference/API_CreateTrainingJob.html).

Reviewer: ```suggestion
2. Train the KNN algorithm on the data using the `SageMakerTrainingOperator`. The configuration for this operator requires information about the algorithm, hyper parameters, the input and output data configuration, resource specifications for the machine running the training job, and the Role ARN for execution. For more information about submitting a training job, check out the [API documentation](https://docs.aws.amazon.com/sagemaker/latest/APIReference/API_CreateTrainingJob.html).
```

Reviewer: Can you put the required information in bullet points like you did above?

3. Create a SageMaker model based on the training results using the `SageMakerModelOperator`. This step creates a model artifact in SageMaker that can be called on demand to provide inferences. The configuration for this operator requires a name for the model, the Role ARN for execution, the image containing the algorithm (in this case the pre-built SageMaker image for KNN), and the S3 path to the model training artifact. For more information on creating a model, check out the API documentation [here](https://docs.aws.amazon.com/sagemaker/latest/APIReference/API_CreateModel.html).

> 4. Evaluate the model on the test data created in Step 1 using the `SageMakerTransformOperator`. This step runs a batch transform to get inferences on the test data from the model created in Step 3. The configuration for this operator requires information about the input data source, the output results path, resource specifications for the machine running the training job, and the name of the model. For more information on submitting a batch transform job, check out the API documentation [here](https://docs.aws.amazon.com/sagemaker/latest/APIReference/API_CreateTransformJob.html).

Reviewer: ```suggestion
4. Evaluate the model on the test data created in Step 1 using the `SageMakerTransformOperator`. This step runs a batch transform to get inferences on the test data from the model created in Step 3. The configuration for this operator requires information about the input data source, the output results path, resource specifications for the machine running the training job, and the name of the model. For more information on submitting a batch transform job, check out the [API documentation](https://docs.aws.amazon.com/sagemaker/latest/APIReference/API_CreateTransformJob.html).
```

Same goes here, best to have these items in a bulleted list.

        aws_conn_id='aws-sagemaker',
        wait_for_completion=True
    )

    create_model = SageMakerModelOperator(
        task_id='create_model',
        config=model_config,
        aws_conn_id='aws-sagemaker'
    )

    test_model = SageMakerTransformOperator(
        task_id='test_model',
        config=transform_config,
        aws_conn_id='aws-sagemaker'
    )

    data_prep >> train_model >> create_model >> test_model
```

And the graph view of the DAG looks like this:

![SageMaker Pipeline](https://assets2.astronomer.io/main/guides/sagemaker/sagemaker_pipeline.png)

When using this DAG, there are a couple of other things to be aware of for your Airflow environment:

- All of the tasks in this DAG make use of a `aws-sagemaker` connection ID to connect to the AWS environment. See the Use Case 1 section above for a description and screenshot of an example connection.
- Some SageMaker operators may require the `AWS_DEFAULT_REGION` to be set in your Airflow environment in addition to a region being specified in the AWS connection. If you're running on Astronomer, you can set this variable in the UI or in your Dockerfile (e.g. `ENV AWS_DEFAULT_REGION=us-east-2`).

> - Many of the SageMaker operators require a Role ARN to be provided in the configuration. If you don't want to store this directly in your DAG file, you can consider storing it as an Airflow variable.

Reviewer: ```suggestion
- Many of the SageMaker operators require a Role ARN to be provided in the configuration. If you don't want to store this directly in your DAG file, consider storing it as an Airflow variable.
```

- Also mentioned in the Use Case 1 section above, some SageMaker operators require XCom pickling to be turned on in order to work because they return objects that are not JSON serializable. To enable XCom pickling, set `AIRFLOW__CORE__ENABLE_XCOM_PICKLING=True`

> This highlights a basic example of how Airflow can be used with SageMaker to automate an end-to-end ML pipeline. A natural next step would be to deploy this model to a SageMaker endpoint using the `SageMakerEndpointConfigOperator` and `SageMakerEndpointOperator`, which provisions resources to host the model. In general, the SageMaker modules in the AWS provider allow for many possibilities when using Airflow to orchestrate ML pipelines so no matter your use case, Airflow has you covered.

Reviewer: ```suggestion
This highlights a basic example of how Airflow can be used with SageMaker to automate an end-to-end ML pipeline. A natural next step would be to deploy this model to a SageMaker endpoint using the `SageMakerEndpointConfigOperator` and `SageMakerEndpointOperator`, which provisions resources to host the model. In general, the SageMaker modules in the AWS provider allow for many possibilities when using Airflow to orchestrate ML pipelines. No matter your use case, Airflow has you covered.
```

For our pipeline, we take data from a local CSV and upload it to S3. Then we make predictions on that data with the SageMaker model by submitting a [SageMaker batch transform job](https://docs.aws.amazon.com/sagemaker/latest/dg/how-it-works-batch.html). Batch transforms are useful when you need to run inferences on large datasets and don't have a persistent endpoint hosted for your model. They require input data to be provided in S3, and will save output results to an S3 path provided. For the last step in our pipeline, we load the results CSV into Redshift.

To implement the complete pipeline, we have to implement the following steps:

1. Load the data from local storage (`include/`) into S3 using the `PythonOperator` with `S3Hook`. 
2. Submit a SageMaker transform job to get inferences on the data using the `SageMakerTransformOperator`. The operator requires a transform job configuration, which in this case is provided in the `transform_config` dictionary at the top of the DAG file. At a minimum, the transform job configuration requires:

>     - A unique name

Reviewer: ```suggestion

    - A unique name
```

I always separate indented bits with newlines on both sides, sometimes it gets wonky otherwise.

## Use Case 2: Orchestrate a Full ML Pipeline

Our second use case is using Airflow to orchestrate an end-to-end ML pipeline in SageMaker. We use Airflow to acquire and pre-process the data, train a model, create a model from the training results, and evaluate the model on test data using a batch transform job. This use case is relevant if you want to automate the model training, testing, and deployment parts of your ML pipeline.

For this example, we use the [Iris dataset](https://archive.ics.uci.edu/ml/datasets/iris), and train a built-in SageMaker K-Nearest Neighbors (KNN) model. The general steps in the DAG are:

1. Using a `PythonOperator`, grab the data from the API, complete some pre-processing so the data is compliant with KNN requirements, split into train and test sets, and save them to S3 using the `S3Hook`.
2. Train the KNN algorithm on the data using the `SageMakerTrainingOperator`. The configuration for this operator requires: 

>     - Information about the algorithm being used

Reviewer: ```suggestion

    - Information about the algorithm being used
```

    - Any required hyper parameters
    - The input data configuration
    - The output data configuration
    - Resource specifications for the machine running the training job 
    - The Role ARN for execution

    For more information about submitting a training job, check out the [API documentation](https://docs.aws.amazon.com/sagemaker/latest/APIReference/API_CreateTrainingJob.html).

3. Create a SageMaker model based on the training results using the `SageMakerModelOperator`. This step creates a model artifact in SageMaker that can be called on demand to provide inferences. The configuration for this operator requires:

>     - A name for the model

Reviewer: ```suggestion

    - A name for the model
```

    - The Role ARN for execution
    - The image containing the algorithm (in this case the pre-built SageMaker image for KNN)
    - The S3 path to the model training artifact 
    
    For more information on creating a model, check out the API documentation [here](https://docs.aws.amazon.com/sagemaker/latest/APIReference/API_CreateModel.html).

> 4. Evaluate the model on the test data created in Step 1 using the `SageMakerTransformOperator`. This step runs a batch transform to get inferences on the test data from the model created in Step 3. The configuration for this operator requires:

Reviewer: ```suggestion
4. Evaluate the model on the test data created in Step 1 using the `SageMakerTransformOperator`. This step runs a batch transform to get inferences on the test data from the model created in Step 3. The configuration for this operator requires:

```
